File: clients/javascript/__tests__/features/ast.spec.js
import {default as ASTClient} from "../../src/clients/ast";

require("../support/acceptance");

const Proc = require("../../index");
const client = Proc.connect(process.env.SECRET, {client: ASTClient});

test("builds the ast for simple calls", async () => {
  expect(await client.core.echo("foo").call()).toEqual([["$$", "proc", ["{}", ["()", "core.echo", [">>", ["%%", "foo"]]]]]]);
});

test("builds the ast for compositions", async () => {
  const composition = client.core.echo("foo").compose(client.core.echo());

  expect(await composition.call()).toEqual(
    [
      [
        "$$", "proc", [
          "{}",
          [">>", ["%%", "foo"]],
          ["()", "core.echo", [">>", ["%%", "foo"]]],
          ["()", "core.echo"]
        ]
      ]
    ]
  );
});

File: clients/ruby/spec/features/ast_spec.rb
# frozen_string_literal: true

require_relative "support/context/acceptance"

require "proc/clients/ast"

RSpec.describe "ruby client api: building asts" do
  include_context "acceptance"

  let(:client_class) {
    Proc::Clients::AST
  }

  it "builds the ast for simple calls" do
    expect(client.core.echo("foo").call).to eq([["$$", "proc", ["{}", ["()", "core.echo", [">>", ["%%", "foo"]]]]]])
  end

  it "builds the ast for compositions" do
    expect((client.core.echo("foo") >> client.core.echo).call).to eq(
      [
        [
          "$$", "proc", [
            "{}",
            [">>", ["%%", "foo"]],
            ["()", "core.echo", [">>", ["%%", "foo"]]],
            ["()", "core.echo"]
          ]
        ]
      ]
    )
  end
end
